### PR TITLE
Add schema.org QAPage metadata

### PR DESCRIPTION
**Pull request by notriddle**
This is an SEO trick, intended to make results from Discourse forms more "clickable" by promising a solution at the other end of the link.

https://meta.discourse.org/t/suggestion-add-checkmark-for-solved-threads-to-the-topic-title-as-well/33813

### CODE REVIEW

**Comment by coding-horror** [comment]
Not a fan of this approach.. why not use the unicode solved emoji and then the slug would naturally contain "solved" because that's the slug for the emoji? Also this will conflict with the visual presentation of solved topics in the topic list, now they will be doubly marked?

**Comment by notriddle** [comment]
> Also this will conflict with the visual presentation of solved topics in the topic list, now they will be doubly marked?

No, page_title is not used in the topic lists. It is, literally, only used in generated HTML.

**Comment by coding-horror** [comment]
hmm, there is some good discussion, what do you think @SamSaffron ? https://meta.discourse.org/t/suggestion-add-checkmark-for-solved-threads-to-the-topic-title-as-well/33813

If it is limited to the HTML part, that seems fine, if we can agree on what it should be.

**Comment by notriddle** [comment]
I'm patching [this function](https://github.com/discourse/discourse/blob/master/lib/topic_view.rb#L147-L157), which also (optionally) adds category information to the titles. If this was going to end up messing up the internals of the page, it would already be duplicating that.

**Comment by coopr** [comment]
Is this going to get merged? I'd like this too!

**Comment by coding-horror** [comment]
@SamSaffron never commented soo 🤷‍♂️

**Comment by SamSaffron** [comment]
I feel like this should be optional https://schema.org/QAPage is a far better metadata format for search title vs hacking up the title

**Comment by notriddle** [comment]
You sure we don't want to use <https://schema.org/Question>?

**Comment by notriddle** [comment]
I agree.

**Comment by notriddle** [comment]
@davidtaylorhq There, now the diff only shows one file changed.

File: plugin.rb [hunk 1]
```diff
@@ -212,6 +212,65 @@ def limit_accepts
     ["is_accepted_answer"]
   end
 
+  register_html_builder('server:head-metadata') do |controller|
```
[the remark this answers is not in the document]
Comment (notriddle) [reply plugin.rb]: @SamSaffron Switch done.

File: plugin.rb [hunk 2]
```diff
@@ -211,6 +211,73 @@ def limit_accepts
...
+    question_json =  {
+      '@type' => 'Question',
+      'name' => topic.title,
+      'text' => first_post.raw,
+      'upvoteCount' => first_post.like_count,
+      'answerCount' => topic.reply_count,
+      'dateCreated' => first_post.created_at,
+      'author' => {
+        '@type' => 'Person',
+        'name' => first_post.user.username
```
Comment (SamSaffron) [inline plugin.rb]: This is a bit risky, can you change all of this to `&.` sometimes posts have no users, we don't want it to explode

File: plugin.rb [hunk 3]
```diff
@@ -211,6 +211,73 @@ def limit_accepts
...
+      }
+    }
+    page_json = {
+      '@type' => 'QAPage',
+      'name' => topic.title,
+    }
+    accepted_answer_post_id = topic.custom_fields["accepted_answer_post_id"]
```
Comment (SamSaffron) [inline plugin.rb]:
```suggestion
    if accepted_answer = Post.find_by(id: topic.custom_fields["accepted_answer_post_id"])
```

safer and more concise

File: plugin.rb [hunk 4]
```diff
@@ -211,6 +211,73 @@ def limit_accepts
...
+        'dateCreated' => accepted_answer.created_at,
+        'url' => accepted_answer.full_url,
+        'author' => {
+          '@type' => 'Person',
+          'name' => accepted_answer.user.username
+        }
+      }
+    end
+    suggested_answer = Post.where(topic_id: topic.id, post_number: controller.params[:post_number]).first
+    if suggested_answer and suggested_answer.post_number != 1
```
Comment (SamSaffron) [inline plugin.rb]:
```suggestion
    if suggested_answer && suggested_answer.post_number != 1
```

and is strange in ruby has weird precedence we always use && in conditionals

File: plugin.rb [hunk 5]
```diff
@@ -211,6 +211,73 @@ def limit_accepts
...
+        'upvoteCount' => accepted_answer.like_count,
+        'dateCreated' => accepted_answer.created_at,
+        'url' => accepted_answer.full_url,
+        'author' => {
+          '@type' => 'Person',
+          'name' => accepted_answer.user.username
+        }
+      }
+    end
+    suggested_answer = Post.where(topic_id: topic.id, post_number: controller.params[:post_number]).first
```
Comment (SamSaffron) [inline plugin.rb]: find_by

**Comment by SamSaffron** [comment]
need a tiny bit more cleanup but I will do the follow up commit, thanks!

**Pull request merged**